homeassistant/components/tradfri/base_class.py

```diff
@@ -28,23 +31,58 @@ def __init__(self, device, api, gateway_id):
 
         self._refresh(device)
 
+    def _restart(self, message, exc):
+        """Log error and restart observation of device."""
+        # Force refresh to handle a situation where the GUI is not updated
+        self.async_schedule_update_ha_state(force_refresh=True)
+
+        _LOGGER.debug(message, exc_info=exc)
+        # Wait one half second before trying again
+        asyncio.sleep(0.5)
+        self._async_start_observe()
+        return True
+
     @callback
     def _async_start_observe(self, exc=None):
         """Start observation of device."""
         if exc:
-            self.async_schedule_update_ha_state()
-            _LOGGER.warning("Observation failed for %s", self._name, exc_info=exc)
-
-        try:
-            cmd = self._device.observe(
-                callback=self._observe_update,
-                err_callback=self._async_start_observe,
-                duration=0,
-            )
-            self.hass.async_create_task(self._api(cmd))
-        except PytradfriError as err:
-            _LOGGER.warning("Observation failed, trying again", exc_info=err)
-            self._async_start_observe()
+            if isinstance(exc, AttributeError):
+                message = (
+                    f"Caught an exception with {self._name}. Restarting observation."
+                )
+            elif isinstance(exc, RequestTimedOut):
+                message = f"Time out error with {self._name}. Restarting observation."
+            else:
+                message = f"Undefined error with {self._name}. Restarting observation."
+
+            self._restart(message, exc)
+            return False
+
+        cmd = self._device.observe(
+            callback=self._observe_update,
+            err_callback=self._async_start_observe,
+            duration=0,
+        )
+        self.hass.async_create_task(self._api(cmd))
+        return True
+
+    @staticmethod
+    def retry_timeout(api, retries=3):
+        """Retry API call when a timeout occurs."""
+
+        @wraps(api)
+        def retry_api(*args, **kwargs):
+            """Retrying API."""
+            for i in range(1, retries + 1):
+                try:
+                    return api(*args, **kwargs)
+                except PytradfriError as err:
+                    _LOGGER.debug("Retrying Tradfri % due to %" % (i, err))
```

Comment:
Please use `%s` operator and pass the variables as arguments.
```py
_LOGGER.debug("Retrying api call %s due to %s", i, err)
```
We don't need to include the module or integration name in the message. That's already part of the logging message by default.